### PR TITLE
OpenXR: Add support for binding modifiers

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Adds support for OpenXR binding modifiers to the action map.

This allows you to add modifiers to inputs such as applying thresholds or offsetting values, etc.
Binding modifiers can add additional input paths or change the type of an input path.

Todos:
- [x] Create infrastructure for binding extension system to the action map
- [x] Make UI changes to the action map UI
- [x] Implement support for haptic triggers
- [x] Implement analog threshold extension (may move this into vendor plugin)
  - [X] implement resource class
  - [x] implement UI
- [x] Implement dpad binding extension
  - [x] implement resource class
  - [x] implement UI

Depends on:
- #97392
- #98163

There is now a demo for this project that can be found here:
https://github.com/godotengine/godot-demo-projects/pull/1137

Some improvements that can be done (possibly in follow up PRs):
- [ ] Should show our modifier button icon in a different color if there are modifiers defined.
- [x] Should find a way to limit the modifiers you can create to those that are applicable.

*Contributed by Khronos Group through the Godot Integration Project*

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
Back working on this after dealing with a bout of the flu. Starting to get an interface together but need to change direction slightly.

**Comment by BastiaanOlij** [comment]
While some of this is a bit of an unknown (but likely) future, I changed our action map structure slightly so we can record binding modifiers both on interaction profiles and on individual bindings.

Right now the dpad modifier is a bit of the odd duckling being the only modifier that is linked to an action set/source path combination and is thus recorded on the interaction profile level.
The analog threshold modifier is linked to an action/source path combo which is recorded as a binding (hence the changes in #98163) and it is very likely that this is going to be true for future modifiers as well.

Right now they are all added to the next chain or `XrInteractionProfileSuggestedBinding` when calling `xrSuggestInteractionProfileBindings` however this may change/be enhanced in the future.

**Comment by BastiaanOlij** [comment]
UI is finally coming together. Binding modifiers are accessible either on the interaction profile (1) or for individual bindings (2):
image

Either button opens up a popup that allows for creating applicable binding modifiers on that level:
image

I still need to change the dialog for selecting a new binding modifier as it doesn't filter by type, and we only have the two modifiers right now but that will change in the near future.

**Comment by BastiaanOlij** [comment]
Some interesting feedback concerning the DPad modifier extension that we need to ensure gets documented well.

The extension actually does two things when enabled.
1. It adds new binding paths you can use for each thumbstick and thumbpad input. These paths are immediately usable even if you do not add a dpad binding modifier
2. And it adds the dpad binding modifier, which lets you customise the dpad behaviour

**Comment by BastiaanOlij** [comment]
All the new paths have been added to the meta data database. In doing this I've cleaned up a lot of the code to make it easier to manage. Still planning on adding a bit of extra logic in the UI when editing the action map so we only show bindings that are available based on which extensions have been selected in project settings so we don't get a whole bunch the user isn't using.

**Comment by BastiaanOlij** [comment]
Other then needing to do some more testing, this should now all work.

My only remaining issue is that I had to use the `EditorInspector` class to make the property editors work properly but these introduce the resource section into the UI and I haven't found a way to supress this:

image

These sections make no sense in this context so I hope there is a way to remove them.

**Comment by BastiaanOlij** [comment]
One more extra thing, this PR also introduces the `OpenXRHapticBase` base class and `OpenXRHapticVibration` class, currently only used for the new features in the action map. There are a number of vendor specific implementations of the haptic feedback OpenXR struct that forms the base of this implementation that are also used for triggering haptic feedback pulses from code.

I'm thinking of replacing `OpenXRAPI::trigger_haptic_pulse` with a version that takes an `OpenXRHapticBase` object as a parameter and we'll be able to perform different haptic effects as they are added to the OpenXR spec generically.
This is something we'll do in a separate PR however.

**Comment by fire** [comment]
Maybe you can figure out why the tests fail.

```

editor/editor_inspector.cpp:4098:57: runtime error: member access within null pointer of type 'struct EditorFeatureProfileManager'
editor/editor_inspector.cpp:4098:154: runtime error: member access within null pointer of type 'struct Node'

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.4.dev.gh-97140 (b26897c0a0f76a2c8ee1f48419e975911b95d691)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(+0x380b26f7) [0x55bd781cc6f7] (??:?)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7fc58e828090] (??:0)
[3] EditorInspector::_notification(int) (??:?)
[4] EditorInspector::_notificationv(int, bool) (??:?)
[5] Object::notification(int, bool) (??:?)
[6] Node::_propagate_ready() (??:?)
[7] Node::_propagate_ready() (??:?)
[8] Node::_propagate_ready() (??:?)
[9] Node::_set_tree(SceneTree*) (??:?)
[10] Node::_add_child_nocheck(Node*, StringName const&, Node::InternalMode) (??:?)
[11] Node::add_child(Node*, bool, Node::InternalMode) (??:?)
[12] void call_with_variant_args_helper<__UnexistingClass, Node*, bool, Node::InternalMode, 0ul, 1ul, 2ul>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, Callable::CallError&, IndexSequence<0ul, 1ul, 2ul>) (??:?)
[13] void call_with_variant_args_dv<__UnexistingClass, Node*, bool, Node::InternalMode>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (??:?)
[14] MethodBindT<Node*, bool, Node::InternalMode>::call(Object*, Variant const**, int, Callable::CallError&) const (??:?)
[15] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (??:?)
[16] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (??:?)
[17] Object::callp(StringName const&, Variant const**, int, Callable::CallError&) (??:?)
[18] Variant::callp(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (??:?)
[19] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (??:?)
[20] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (??:?)
[21] Node::_gdvirtual__ready_call() (??:?)
[22] Node::_notification(int) (??:?)
[23] Node::_notificationv(int, bool) (??:?)
[24] Object::notification(int, bool) (??:?)
[25] Node::_propagate_ready() (??:?)
[26] Node::_set_tree(SceneTree*) (??:?)
[27] Node::_add_child_nocheck(Node*, StringName const&, Node::InternalMode) (??:?)
[28] Node::add_child(Node*, bool, Node::InternalMode) (??:?)
[29] void call_with_variant_args_helper<__UnexistingClass, Node*, bool, Node::InternalMode, 0ul, 1ul, 2ul>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, Callable::CallError&, IndexSequence<0ul, 1ul, 2ul>) (??:?)
[30] void call_with_variant_args_dv<__UnexistingClass, Node*, bool, Node::InternalMode>(__UnexistingClass*, void (__UnexistingClass::*)(Node*, bool, Node::InternalMode), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (??:?)
[31] MethodBindT<Node*, bool, Node::InternalMode>::call(Object*, Variant const**, int, Callable::CallError&) const (??:?)
[32] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (??:?)
[33] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (??:?)
[34] Node::_gdvirtual__process_call(double) (??:?)
[35] Node::_notification(int) (??:?)
[36] Node::_notificationv(int, bool) (??:?)
[37] CanvasItem::_notificationv(int, bool) (??:?)
[38] Control::_notificationv(int, bool) (??:?)
[39] Object::notification(int, bool) (??:?)
[40] SceneTree::_process_group(SceneTree::ProcessGroup*, bool) (??:?)
[41] SceneTree::_process(bool) (??:?)
[42] SceneTree::process(double) (??:?)
[43] Main::iteration() (??:?)
[44] OS_LinuxBSD::run() (??:?)
[45] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(main+0x4a3) [0x55bd781cbf9c] (??:?)
[46] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fc58e809083] (??:0)
[47] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(_start+0x2e) [0x55bd781cba3e] (??:?)
-- END OF BACKTRACE --
```

**Comment by BastiaanOlij** [comment]
> Maybe you can figure out why the tests fail.

Looks like a failure to check for NULL parameters with standard unit tests. Sadly on a part of the API that is completely unrelated to this work, so I don't know why it's suddenly failing.

**Comment by BastiaanOlij** [comment]
Hmm I wonder, the line of code that fails is `EditorFeatureProfileManager::get_singleton()->connect("current_feature_profile_changed", callable_mp(this, &EditorInspector::_feature_profile_changed));`
I'm guessing the testing doesn't setup the editor classes so possibly me using the EditorInspector in classes that are exposed as they are designed to be subclassed in the vendors plugin.

**Review by dsnopek — COMMENTED** [review]
Thanks!

This PR is epic in size, but it looks great! All my code review comments are very nitpick-y things.

I tested the new editor stuff, and it all seemed to work fine. I didn't attempt to test it functionally - I see that SteamVR and Monado reportedly support it, so I may attempt to give it a try later.

Since PR https://github.com/godotengine/godot/pull/98163 has been merged, this should be rebased. That'll make reviewing a tiny bit easier.

File: modules/openxr/action_map/openxr_action_map.cpp [hunk 1]
```diff
@@ -107,14 +107,26 @@ void OpenXRActionMap::remove_action_set(Ref<OpenXRActionSet> p_action_set) {
 	}
 }
 
-void OpenXRActionMap::set_interaction_profiles(Array p_interaction_profiles) {
+void OpenXRActionMap::clear_interaction_profiles() {
+	// Interaction profiles held within our action map set should be released and destroyed but just in case they are still used some where else
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_action_map.cpp:111]: The comment was more related to the function as a whole, but yeah maybe clearer if I move it to the loop.

File: modules/openxr/editor/openxr_binding_modifier_editor.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// OpenXRBindingModifierEditor
+
+void OpenXRBindingModifierEditor::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_binding_modifier"), &OpenXRBindingModifierEditor::get_binding_modifier);
+	ClassDB::bind_method(D_METHOD("setup", "action_map", "binding_modifier"), &OpenXRBindingModifierEditor::setup);
+
+	ClassDB::bind_method("_on_remove_binding_modifier", &OpenXRBindingModifierEditor::_on_remove_binding_modifier);
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/editor/openxr_binding_modifier_editor.cpp:217]: I think this is something I mostly do out of habit because once upon a time, this was needed.

[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/editor/openxr_binding_modifier_editor.cpp:217]: I could remove it but we should really clear out a whole bunch of em.

File: editor/editor_inspector.cpp [hunk 3]
```diff
@@ -4095,6 +4095,7 @@ void EditorInspector::_notification(int p_what) {
 		} break;
 
 		case NOTIFICATION_READY: {
+			ERR_FAIL_NULL(EditorFeatureProfileManager::get_singleton());
```
Comment (BastiaanOlij) [inline editor/editor_inspector.cpp:4098]: Now that CI testing runs through this, and we're not fully setting up the editor, we need to exit here as the `EditorFeatureProfileManager` singleton does not get constructed.

File: modules/openxr/doc_classes/OpenXRBindingModifier.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+			<description>
+				Returns the data that is sent to OpenXR when submitting the suggested interacting bindings this modifier is a part of.
+				[b]Note:[/b] This must be data compatible with a [code]XrBindingModificationBaseHeaderKHR[/code] structure.
+			</description>
+		</method>
+		<method name="_record_on_binding" qualifiers="virtual const">
+			<return type="bool" />
+			<description>
+				Return [code]false[/code] if binding modifiers of this type are recorded on an interaction profile.
+				Return [code]true[/code] if binding modifiers of this type are recorded on bindings within an interaction profile.
```
Comment (m4gr3d) [inline modules/openxr/doc_classes/OpenXRBindingModifier.xml:29]: The description is a bit confusing.

File: modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		</member>
+		<member name="center_region" type="float" setter="set_center_region" getter="get_center_region" default="0.1">
+			Center region in which our center position of our dpad return [code]true[/code].
+		</member>
+		<member name="input_path" type="String" setter="set_input_path" getter="get_input_path" default="&quot;&quot;">
+			Input path for this dpad binding modifier.
+		</member>
+		<member name="is_sticky" type="bool" setter="set_is_sticky" getter="get_is_sticky" default="false">
+			If [code]false[/code], when the joystick enters a new dpad zone this becomes true.
+			If [code]true[/code], when the joystick remains in active dpad zone, this remains true even if we overlap with another zone.
```
Comment (m4gr3d) [inline modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml:25]: Also hard to understand what this field is for at first read.

File: modules/openxr/editor/openxr_action_map_editor.cpp [hunk 6]
```diff
@@ -387,6 +415,22 @@ void OpenXRActionMapEditor::_clear_action_map() {
...
+void OpenXRActionMapEditor::register_binding_modifier_editor(const String &p_binding_modifier_class, const String &p_editor_class) {
+	binding_modifier_editors[p_binding_modifier_class] = p_editor_class;
+}
+
+String OpenXRActionMapEditor::get_binding_modifier_editor_class(const String &p_binding_modifier_class) {
+	if (binding_modifier_editors.has(p_binding_modifier_class)) {
+		return binding_modifier_editors[p_binding_modifier_class];
+	}
+
+	return "OpenXRBindingModifierEditor";
```
Comment (m4gr3d) [inline modules/openxr/editor/openxr_action_map_editor.cpp:431]: How about `OpenXRBindingModifierEditor::get_class_name()` to avoid the use of magic strings and provide compilation safety.

File: modules/openxr/editor/openxr_binding_modifier_editor.cpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+bool EditorInspectorPluginBindingModifier::can_handle(Object *p_object) {
+	Ref<OpenXRBindingModifier> binding_modifier(Object::cast_to<OpenXRBindingModifier>(p_object));
+	return binding_modifier.is_valid();
+}
+
+bool EditorInspectorPluginBindingModifier::parse_property(Object *p_object, const Variant::Type p_type, const String &p_path, const PropertyHint p_hint, const String &p_hint_text, const BitField<PropertyUsageFlags> p_usage, const bool p_wide) {
+	Ref<OpenXRBindingModifier> binding_modifier(Object::cast_to<OpenXRBindingModifier>(p_object));
+	ERR_FAIL_COND_V(binding_modifier.is_null(), false);
+
+	if (p_type == Variant::OBJECT && p_hint == PROPERTY_HINT_RESOURCE_TYPE && p_hint_text == "OpenXRActionSet") {
```
Comment (m4gr3d) [inline modules/openxr/editor/openxr_binding_modifier_editor.cpp:172]: Similar comment; can we use `p_hint_text == OpenXRActionSet::get_class_name()`?

File: modules/openxr/action_map/openxr_interaction_profile.cpp [hunk 8]
```diff
@@ -81,6 +87,76 @@ String OpenXRIPBinding::get_binding_path() const {
...
+	}
+}
+
+Array OpenXRIPBinding::get_binding_modifiers() const {
+	Array ret;
+	for (const Ref<OpenXRBindingModifier> &binding_modifier : binding_modifiers) {
+		ret.push_back(binding_modifier);
+	}
+	return ret;
+}
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_interaction_profile.cpp:130]: Because we're converting from a `Vector<Ref<OpenXRBindingModifier>>` to an `Array`. I don't think there is an implicit conversion or was that added at some time?

File: modules/openxr/action_map/openxr_action_map.cpp [hunk 9]
```diff
@@ -156,6 +175,10 @@ void OpenXRActionMap::remove_interaction_profile(Ref<OpenXRInteractionProfile> p
 	int idx = interaction_profiles.find(p_interaction_profile);
 	if (idx != -1) {
 		interaction_profiles.remove_at(idx);
+
+		ERR_FAIL_COND_MSG(p_interaction_profile->action_map != this, "Removing interaction profile that belongs to this action map but had incorrect action map pointer."); // this should never happen!
+		p_interaction_profile->action_map = nullptr;
```
[the remark this answers is not in the document]
Comment (BastiaanOlij) [reply modules/openxr/action_map/openxr_action_map.cpp:180]: No, the design is that an interaction profile always belongs to a single action map. The `action_map` member is only null momentarily in 2 conditions:
1. when we're creating a new interaction profile right before it's added to an action map
2. when we're destructing an interaction profile where it's remove from the action map just before.

The check to see if the actionmap pointer is different is just to catch errors in usage, someone who later on adds code that would attempt to add an interaction profile to multiple action maps, and these checks just erroring out.

File: modules/openxr/doc_classes/OpenXRBindingModifier.xml (hunk 4 above)
Comment (BastiaanOlij) [reply modules/openxr/doc_classes/OpenXRBindingModifier.xml:29]: Basically there are two types of binding modifiers, those recorded on the interaction profile, and those recorded against a specific binding on the interaction profile. This flag simply indicates which type we're dealing with.
This really won't get a lot cleared until more binding modifiers become public.

File: modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml (hunk 5 above)
Comment (BastiaanOlij) [reply modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml:25]: I had real difficulty describing this in text, you really need to look at the diagrams in the API spec and even there the description is overly technical.

Comment (m4gr3d) [reply modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml:25]: Are those diagrams publicly available? Can you link to them in this description?

**Review by m4gr3d — COMMENTED** [review]
Logic looks good, although I'd like us to iterate on the documentation as in their current state, they're hard to parse.

@BastiaanOlij Is there a test project that can be used to help validate those changes?

**Comment by BastiaanOlij** [comment]
> @BastiaanOlij Is there a test project that can be used to help validate those changes?

@m4gr3d, I added a demo project on godot-demo-projects, see link in OP.

File: modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml (hunk 5 above)
Comment (BastiaanOlij) [reply modules/openxr/doc_classes/OpenXRDpadBindingModifier.xml:25]: If you scroll to the top of the help page you'll notice that on both modifiers, I link to the specification for the modifiers.

For the DPad modifier its this one: https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_EXT_dpad_binding

Totally open to suggestions on how to work that more user friendly and embed that into our documentation.

File: modules/openxr/openxr_api.cpp [hunk 10]
```diff
@@ -513,40 +513,66 @@ void OpenXRAPI::copy_string_to_char_buffer(const String p_string, char *p_buffer
...
+	return requested_extensions;
+}
+
 bool OpenXRAPI::create_instance() {
 	// Create our OpenXR instance, this will query any registered extension wrappers for extensions we need to enable.
 
-	// Append the extensions requested by the registered extension wrappers.
-	HashMap<String, bool *> requested_extensions;
+	// We can request an extension multiple times if there are dependencies
+	struct request_extension {
```
Comment (dsnopek) [inline modules/openxr/openxr_api.cpp:538]: Even though this is a very local struct, I think we always use camel case for struct names - using snake case feels a little weird

File: modules/openxr/action_map/openxr_binding_modifier.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+private:
+protected:
+	friend class OpenXRInteractionProfile;
+	friend class OpenXRIPBinding;
+
+	static void _bind_methods();
+
+	GDVIRTUAL0RC(bool, _record_on_binding)
+	GDVIRTUAL0RC(String, _get_description)
+	GDVIRTUAL0R(PackedByteArray, _get_ip_modification)
```
Comment (dsnopek) [inline modules/openxr/action_map/openxr_binding_modifier.h:55]: These feel like they should be required? So:

```suggestion
	GDVIRTUAL0RC_REQUIRED(bool, _record_on_binding)
	GDVIRTUAL0RC_REQUIRED(String, _get_description)
	GDVIRTUAL0R_REQUIRED(PackedByteArray, _get_ip_modification)
```
This will make it an error if `GDVIRTUAL_CALL()` can't find the method, and mark them as required in the `extension_api.json`.

This won't have any effect on the sub-classes within Godot that override these methods; only for GDExtension classes that do.

File: modules/openxr/action_map/openxr_binding_modifier.h [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+// Part of implementation for:
+// https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XR_KHR_binding_modification
+
+class OpenXRInteractionProfile;
+class OpenXRIPBinding;
+
+class OpenXRBindingModifier : public Resource {
+	GDCLASS(OpenXRBindingModifier, Resource);
+
+private:
```
Comment (dsnopek) [inline modules/openxr/action_map/openxr_binding_modifier.h:46]: This superfluous `private:` could be removed:

```suggestion
```

File: modules/openxr/editor/openxr_binding_modifier_editor.h [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+/**************************************************************************/
+
+#ifndef OPENXR_BINDING_MODIFIER_EDITOR_H
+#define OPENXR_BINDING_MODIFIER_EDITOR_H
+
+#include "../action_map/openxr_action_map.h"
+#include "../action_map/openxr_action_set.h"
+#include "../action_map/openxr_binding_modifier.h"
+#include "editor/editor_inspector.h"
+// #include "editor/editor_properties.h"
```
Comment (dsnopek) [inline modules/openxr/editor/openxr_binding_modifier_editor.h:38]: Delete this?

```suggestion
```

File: modules/openxr/extensions/openxr_valve_analog_threshold_extension.h [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+public:
+	OpenXRAnalogThresholdModifier();
+
+	void set_on_threshold(float p_threshold);
+	float get_on_threshold() const;
+
+	void set_off_threshold(float p_threshold);
+	float get_off_threshold() const;
+
+	void set_on_haptic(const Ref<OpenXRHapticBase> p_haptic);
```
Comment (dsnopek) [inline modules/openxr/extensions/openxr_valve_analog_threshold_extension.h:78]: I think this might have meant to be a const reference? So:
```suggestion
	void set_on_haptic(const Ref<OpenXRHapticBase> &p_haptic);
```

File: modules/openxr/extensions/openxr_valve_analog_threshold_extension.h [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	Ref<OpenXRHapticBase> get_on_haptic() const;
+
+	void set_off_haptic(const Ref<OpenXRHapticBase> p_haptic);
```
Comment (dsnopek) [inline modules/openxr/extensions/openxr_valve_analog_threshold_extension.h:81]: This one too:
```suggestion
	void set_off_haptic(const Ref<OpenXRHapticBase> &p_haptic);
```

File: modules/openxr/extensions/openxr_dpad_binding_extension.h [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+	void set_wedge_angle(float p_wedge_angle);
+	float get_wedge_angle() const;
+
+	void set_wedge_angle_deg(float p_wedge_angle);
+	float get_wedge_angle_deg() const;
+
+	void set_is_sticky(bool p_sticky);
+	bool get_is_sticky() const;
+
+	void set_on_haptic(const Ref<OpenXRHapticBase> p_haptic);
```
Comment (dsnopek) [inline modules/openxr/extensions/openxr_dpad_binding_extension.h:99]: And this:
```suggestion
	void set_on_haptic(const Ref<OpenXRHapticBase> &p_haptic);
```

File: modules/openxr/extensions/openxr_dpad_binding_extension.h [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	Ref<OpenXRHapticBase> get_on_haptic() const;
+
+	void set_off_haptic(const Ref<OpenXRHapticBase> p_haptic);
```
Comment (dsnopek) [inline modules/openxr/extensions/openxr_dpad_binding_extension.h:102]: And here:
```suggestion
	void set_off_haptic(const Ref<OpenXRHapticBase> &p_haptic);
```

**Review by dsnopek — COMMENTED** [review]
I skimmed the latest code, and mostly just found more little nitpicky stuff.

I haven't tested the binding modifiers at runtime, but I did re-test the editor stuff again. I'm able to add an `OpenXRAnalogThresholdModifier`, and that seems to be working fine. However, when I try to add an `OpenXRDpadBindingModifier` I get this error:

```
ERROR: Condition "!new_binding_modifier->record_on_binding()" is true.
   at: _on_dialog_created (modules/openxr/editor/openxr_binding_modifiers_dialog.cpp:134)
```

Is that expected? If so, how should dpad modifiers be added?

File: doc/classes/ProjectSettings.xml [hunk 18]
```diff
@@ -2980,6 +2980,12 @@
 		<member name="threading/worker_pool/max_threads" type="int" setter="" getter="" default="-1">
 			Maximum number of threads to be used by [WorkerThreadPool]. Value of [code]-1[/code] means no limit.
 		</member>
+		<member name="xr/openxr/binding_modifiers/analog_threshold" type="bool" setter="" getter="" default="false">
+			Enables the analog threshold binding modifier if supported by the XR runtime.
+		</member>
+		<member name="xr/openxr/binding_modifiers/dpad_binding" type="bool" setter="" getter="" default="false">
+			Enables the dpad binding modifier if supported by the XR runtime.
```
Comment (Mickeon) [inline doc/classes/ProjectSettings.xml:2987]:
```suggestion
			If [code]true[/code], enables the D-pad binding modifier if supported by the XR runtime.
```

File: doc/classes/ProjectSettings.xml [hunk 19]
```diff
@@ -2980,6 +2980,12 @@
 		<member name="threading/worker_pool/max_threads" type="int" setter="" getter="" default="-1">
 			Maximum number of threads to be used by [WorkerThreadPool]. Value of [code]-1[/code] means no limit.
 		</member>
+		<member name="xr/openxr/binding_modifiers/analog_threshold" type="bool" setter="" getter="" default="false">
+			Enables the analog threshold binding modifier if supported by the XR runtime.
```
Comment (Mickeon) [inline doc/classes/ProjectSettings.xml:2984]:
```suggestion
			If [code]true[/code], enables the analog threshold binding modifier if supported by the XR runtime.
```

**Comment by BastiaanOlij** [comment]
> Is that expected? If so, how should dpad modifiers be added?

DPad modifiers aren't added to individual bindings, they are added to the interaction profile with the binding button on the right.
I need to figure out if I can add additional filters to the standard popup. I didn't want to create two separate base classes but maybe i need to go down that route.

**Comment by dsnopek** [comment]
@BastiaanOlij:

> DPad modifiers aren't added to individual bindings, they are added to the interaction profile with the binding button on the right.
> I need to figure out if I can add additional filters to the standard popup. I didn't want to create two separate base classes but maybe i need to go down that route.

I think even just having a more user-friendly error message would be OK. Right now, I'm getting a very programmer-y error and the modifier just fails to be added.

Maybe we just need a popup with an explanation?

**Comment by BastiaanOlij** [comment]
> @BastiaanOlij:
> 
> > DPad modifiers aren't added to individual bindings, they are added to the interaction profile with the binding button on the right.
> > I need to figure out if I can add additional filters to the standard popup. I didn't want to create two separate base classes but maybe i need to go down that route.
> 
> I think even just having a more user-friendly error message would be OK. Right now, I'm getting a very programmer-y error and the modifier just fails to be added.
> 
> Maybe we just need a popup with an explanation?

I decided to go down the subclass path after all. So there is now an `OpenXRIPBindingModifier` class for binding modifiers that are recorded on interaction profile level, and a `OpenXRActionBindingModifier` class for binding modifiers that are recorded on individual actions/bindings.

Right now it seems a little overkill but once more binding modifiers become available it will start to make a whole lot more sense I recon.

**Review by dsnopek — APPROVED** [review]
Thanks!

I re-tested in the editor, adding both threshold and dpad modifiers with haptics in the right places for those. Having the two base classes seems like a nice way to go, both for the editor and making the API simpler (removing the somewhat confusing `record_on_binding` virtual method).

I skimmed the code again and it looks good to me, but it's a big PR and having reviewed it so many times already, my brain can't pick up on all that much anymore. :-)

I think this is ready to go!

File: modules/openxr/editor/openxr_action_map_editor.h [hunk 20]
```diff
@@ -48,6 +49,9 @@ class OpenXRActionMapEditor : public VBoxContainer {
 	GDCLASS(OpenXRActionMapEditor, VBoxContainer);
 
 private:
+	static HashMap<String, String> interaction_profile_editors; // interaction profile path, interaction profile editor
+	static HashMap<String, String> binding_modifier_editors; // binding modifier class, binding modifiers editor
```
Comment (AThousandShips) [inline modules/openxr/editor/openxr_action_map_editor.h:53]:
```suggestion
	static HashMap<String, String> interaction_profile_editors; // interaction profile path, interaction profile editor.
	static HashMap<String, String> binding_modifier_editors; // binding modifier class, binding modifiers editor.
```

File: modules/openxr/editor/openxr_binding_modifiers_dialog.cpp [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+	if (ip_binding.is_valid()) {
+		new_binding_modifiers = ip_binding->get_binding_modifiers();
+	} else if (interaction_profile.is_valid()) {
+		new_binding_modifiers = interaction_profile->get_binding_modifiers();
+	} else {
+		ERR_FAIL_MSG("No binding nor interaction profile specified.");
+	}
+
+	for (int i = 0; i < new_binding_modifiers.size(); i++) {
+		Ref<OpenXRBindingModifier> binding_modifier = new_binding_modifiers[i];
```
Comment (AThousandShips) [inline modules/openxr/editor/openxr_binding_modifiers_dialog.cpp:96]:
```suggestion
	for (Ref<OpenXRBindingModifier> binding_modifier : new_binding_modifiers) {
```

**Comment by BastiaanOlij** [comment]
@dsnopek Thanks for checking it again though, I do think it's time we merge this and any further things we find we can deal with later. It's a big PR but for people who don't use binding modifiers it's fairly low risk.

@AThousandShips dang :) That's a lot of cleanup of existing code. Worth doing I guess, I'll go through it later today :)

**Comment by AThousandShips** [comment]
Limited it largely to new code 😅
